src/System.Runtime.Extensions/src/System/Net/WebUtility.cs

```diff
@@ -367,9 +370,37 @@ public static string UrlEncode(string value)
             if (string.IsNullOrEmpty(value))
                 return value;
 
-            byte[] bytes = Encoding.UTF8.GetBytes(value);
-            byte[] encodedBytes = UrlEncode(bytes, 0, bytes.Length, false /* alwaysCreateNewReturnValue */);
-            return Encoding.UTF8.GetString(encodedBytes, 0, encodedBytes.Length);
+            int cSafe = 0;
+            int cSpaces = 0;
+            for (int i = 0; i < value.Length; i++)
+            {
+                char ch = value[i];
+                if (ch == ' ')
+                {
+                    cSpaces++;
+                }
+                else if (IsUrlSafeChar(ch))
+                {
+                    cSafe++;
+                }
+            }
+
+            if (cSpaces == 0 && cSafe == value.Length)
+            {
+                // Nothing to expand
+                return value;
+            }
+
+            int unsafeByteCount = Encoding.UTF8.GetByteCount(value) - cSafe - cSpaces;
+            int byteIndex = unsafeByteCount * 2;
+            
+            // For perf, we avoid allocating a new array for the encoded bytes, and
+            // instead encode in place
+            byte[] newBytes = new byte[cSafe + cSpaces + unsafeByteCount + byteIndex];
+            Encoding.UTF8.GetBytes(value, 0, value.Length, newBytes, byteIndex);
```

Comment:
This doesn't look right.  UTF8 can use 1-4 bytes per character.  newBytes has at most 3 bytes per character.  Is there going to be enough room for the worst case?  Plus, I don't understand why byteIndex is being passed in as the offset into newBytes.